### PR TITLE
Add code coverage

### PR DESCRIPTION
**Pull request by kingbuzzman**
When working on the PR "[Implements bulk_create for create_batch if available](https://github.com/FactoryBoy/factory_boy/pull/925)" I found the need to ensure that all the code I wrote was properly tested/covered under tests. Started experimenting and came up with this PR to add coverage when using the CI

### CODE REVIEW

**Comment by kingbuzzman** [comment]
@francoisfreitag @rbarrois What do you guys think about this? .. Also it would be nice if you create a key over at [codecov](https://about.codecov.io/) to upload all the files there and not have to download them :/

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -10,22 +10,24 @@ concurrency:
...
       fail-fast: false
       matrix:
         python-version:
-          - "3.7"
-          - "3.8"
-          - "3.9"
-          - "3.10"
-          - "pypy-3.7"
-          - "pypy-3.8"
+          - ["3.7", "py37"]
```
Comment (kingbuzzman) [inline .github/workflows/test.yml:20]: Adds pretty names.. for the coverage filename

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -58,3 +57,16 @@ jobs:
...
+
+      - name: Coverage report
+        if: matrix.run-coverage == 'yes'
+        run: tox -e cov-report
+
+      - name: 'Upload coverage'
+        if: matrix.run-coverage == 'yes'
+        uses: actions/upload-artifact@v3
+        with:
+          name: factory_coverage_${{ matrix.python-version[1] }}_${{ matrix.database-type }}
```
Comment (kingbuzzman) [inline .github/workflows/test.yml:70]: Would be nice to move this over to a codecov / equivalent and not have to download it..

File: Makefile [hunk 3]
```diff
@@ -80,14 +81,19 @@ lint:
 	$(ISORT) --check-only --diff $(EXAMPLES_DIR) $(PACKAGE) $(SETUP_PY) $(TESTS_DIR)
 	check-manifest
 
-coverage:
+coverage-clean:
 	$(COVERAGE) erase
-	$(COVERAGE) run "--include=$(PACKAGE)/*.py,$(TESTS_DIR)/*.py" --branch $(SETUP_PY) test
```
Comment (kingbuzzman) [inline Makefile:85]: This is no longer supported, plus it was broken. `python setup.py test` failed when it encountered `raise unittest.SkipTest(..)`, I simply migrated the code to reuse the same way `make test` works currently

File: setup.cfg [hunk 4]
```diff
@@ -80,3 +81,19 @@ include_trailing_comma = True
 force_grid_wrap = 0
 use_parentheses = True
 line_length = 88
+
+[coverage:run]
+dynamic_context = test_function
+include=
```
Comment (kingbuzzman) [inline setup.cfg:87]: This comes from the `$(COVERAGE) run "--include=$(PACKAGE)/*.py,$(TESTS_DIR)/*.py"...` in the `Makefile`

**Review by kingbuzzman — COMMENTED** [review]
LMKWYT

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -10,22 +10,25 @@ concurrency:
...
-          - "3.10"
-          - "pypy-3.7"
-          - "pypy-3.8"
+          - ["3.7", "py37"]
+          - ["3.8", "py38"]
+          - ["3.9", "py39"]
+          - ["3.10", "py310"]
+          - ["pypy-3.7", "pypy37"]
+          - ["pypy-3.8", "pypy38"]
+          - ["pypy-3.9", "pypy39"]
```
Comment (kingbuzzman) [inline .github/workflows/test.yml:26]: Noticed pypy 3.9 is out and wasn't being tested. Figured it was an oversight?

**Comment by francoisfreitag** [comment]
I’m -0 on running coverage with every run of the CI (once per combination in the python / DB matrix IIUC).
I personally seldom look at the coverage. I agree it is useful on occasions, but just having the recipe to generate the coverage is enough for me. The rest of the time, it’s making builds slowers and taking up online storage. I can be convinced, but initially don’t think the change is warranted.

The various improvements you’ve suggested are welcome, some where already split out in separate PRs, and the coverage context helps tracking what tests exercise a bit of code. It’ll will soon be pulled out in another PR. :+1:

**Comment by francoisfreitag** [comment]
@rbarrois Any thoughts on:
1. measuring coverage systematically when the project CI runs?
2. upload it to a specialized platform (e.g. codecov) to make it easier to access?

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -10,23 +10,25 @@ concurrency:
...
         database-type:
           - "sqlite"
           - "postgres"
+        run-coverage:
+          - "${{ (github.ref == 'refs/heads/master' || contains(github.event.commits[0].message, 'coverage')) && 'yes' || 'no' }}"
```
Comment (kingbuzzman) [inline .github/workflows/test.yml:31]: How about this?

* `master` takes [3:10s](https://github.com/FactoryBoy/factory_boy/actions/runs/2578141014), [3:45s](https://github.com/FactoryBoy/factory_boy/actions/runs/2575372482), [3:50s](https://github.com/FactoryBoy/factory_boy/actions/runs/2578163487) to run without coverage
* this branch without coverage takes [4:13s](https://github.com/FactoryBoy/factory_boy/actions/runs/2579118949) but I suspect it's just github action runners just being a little slower just cause. and with coverage [5:26s](https://github.com/kingbuzzman/factory_boy/actions/runs/2579194000)

The idea is any branch you're working on doesn't run coverage unless it's master/main, and if you want coverage in your local branch to see whats going on.. you can add the string "`coverage`" to your commit message

File: Makefile (hunk 3 above)
Comment (kingbuzzman) [reply Makefile]: Already addressed.

File: setup.cfg [hunk 7]
```diff
@@ -47,6 +47,7 @@ dev =
     isort
     Pillow
     SQLAlchemy
+    sqlalchemy_utils
```
[the remark this answers is not in the document]
Comment (kingbuzzman) [reply setup.cfg]: Already addressed.

File: setup.cfg (hunk 4 above)
Comment (kingbuzzman) [reply setup.cfg:87]: Already addressed.

File: Makefile [hunk 8]
```diff
@@ -80,14 +81,19 @@ lint:
...
-coverage:
+coverage-clean:
 	$(COVERAGE) erase
-	$(COVERAGE) run "--include=$(PACKAGE)/*.py,$(TESTS_DIR)/*.py" --branch -m unittest
-	$(COVERAGE) report "--include=$(PACKAGE)/*.py,$(TESTS_DIR)/*.py"
-	$(COVERAGE) html "--include=$(PACKAGE)/*.py,$(TESTS_DIR)/*.py"
+	@rm -rf .coverage htmlcov/
+
+coverage-test:
+	$(PYTHON_ERROR_ON_WARN) $(COVERAGE_PATH) run -m unittest
```
Comment (kingbuzzman) [inline Makefile:89]: [Continuing the conversation](https://github.com/FactoryBoy/factory_boy/pull/952/files#r908786905)

File: .github/workflows/test.yml (hunk 6 above)
Comment (kingbuzzman) [reply .github/workflows/test.yml]: This has been removed

File: setup.cfg [hunk 9]
```diff
@@ -80,3 +81,19 @@ include_trailing_comma = True
 force_grid_wrap = 0
 use_parentheses = True
 line_length = 88
+
+[coverage:run]
+dynamic_context = test_function
```
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply setup.cfg:86]: #956

File: setup.cfg [hunk 10]
```diff
@@ -80,3 +81,19 @@ include_trailing_comma = True
...
+omit=tests/djapp/migrations*
+
+[coverage:report]
+include=
+    factory/*.py
+    tests/*.py
+omit=tests/djapp/migrations*
+
+[coverage:html]
+show_contexts = True
```
[the remark this answers is not in the document]
Comment (francoisfreitag) [reply setup.cfg:94]: #956

File: Makefile (hunk 8 above)
Comment (kingbuzzman) [reply Makefile]: @rbarrois

Comment (kingbuzzman) [reply Makefile]: @rbarrois friendly bump :)

**Comment by kingbuzzman** [comment]
There is no point having this around

**Pull request closed without merging**